### ISSUE TITLE
Which yields should be produced at the colony plot?

### ISSUE THREAD

**Issue by Nightinggale** [We-the-People-civ4col-mod/Mod#138]
CvPlot::calculatePotentialCityYield sets some yields to always have 0 production, hence disabled for colony plots. The current code:

```cpp
	if (eYield == YIELD_FOOD || eYield == YIELD_LUMBER || eYield == YIELD_STONE || eYield == YIELD_HEMP)
	{
		return 0;
	}

	CvYieldInfo &kYieldInfo = GC.getYieldInfo(eYield);

	if (!kYieldInfo.isCargo() || kYieldInfo.isLivestock())
	{
		return 0;
	}
```

Livestock is any of the yields, which requires some yields already to produce (horses, sheep etc)

The question is which yields should be enabled? Also how to set it up? Since it reads xml data anyway I vote for adding a bool to xml data and rely exclusively on that one. The less hardcoding in the DLL the better as it makes it more obvious for future modders how to modify something.

**Comment by abmpicoli** [comment]
At this code, right? https://github.com/We-the-People-civ4col-mod/Mod/blob/32a8d1a5360b9071d7898ece97b69d053e9c23c0/Project%20Files/DLLSources/CvPlot.cpp#L6068-L6100 ... 

Hey, why disable HEMP, STONE and LUMBER?  FOOD i think it is ok, since it is granted always. 

For the sake of moddability, maybe it should be provide an iYieldInfo.isFood() function, and have that set into xml... For example, I don't see why people can't consume barley, grapes, beer or wine directly if they are hungry...

**Comment by abmpicoli** [comment]
Also, we could set a isMandatoryCityPlotYield function that would say which yields are provided automatically... based on xml flags.

**Comment by Nightinggale** [comment]
`
Hey, why disable HEMP, STONE and LUMBER? FOOD i think it is ok, since it is granted always.
`

That's why I'm asking. I saw this was written in the code. Since I needed a way to make GUI and actual code agree on the yield production at the colony plot, I moved the code to a standalone function and rewrote the checks to be more readable. Other than that I didn't change anything. Since it's gameplay design, I decided it should be a discussion rather than just something I changed without telling anybody.

We could possibly not block food as it will give you the option to switch between something to export and more food. There could be a balance issue in total food production if we follow that path though.

I can kind of see the point in not making the colony plot a huge cattle producing plot. However I think if we are to do something, we should allow all of them, but decrease the production as the number of houses and citizens increase. This way when it's just a small startup colony, people grow stuff in the backyard. In the big city, not so much, but then the player will have plenty of citizens to work other plots.

`
Also, we could set a isMandatoryCityPlotYield function that would say which yields are provided automatically... based on xml flags.
`

I don't quite get what you want this flag to do. The current code will provide any yield provided by the plot unless it's one of the yields being forced to 0 by that function.

`
For the sake of moddability, maybe it should be provide an iYieldInfo.isFood() function, and have that set into xml... For example, I don't see why people can't consume barley, grapes, beer or wine directly if they are hungry...
`

It's not a bad idea to have multiple types of food. M:C has the concept of luxury food, which when grown will provide a different kind of growth unit. However this is off topic related to the colony plot and deserves a new ticket.

**Comment by abmpicoli** [comment]
The isMandatoryCityPlotYield function would return yields that are always produced at the city plot . That is, right now, food. So the if would become , instead of 
```
if (iYield == FOOD) return 0
```
become 
```
if (isMandatoryCityPlotYield(iYield) return 0;
```

**Comment by Nightinggale** [comment]
The naming confused me. I think it should be more like:
```cpp
if (kYieldInfo.isDisableColonyPlotProduction()) return 0;
```

I think we think of the same thing, but it's a question of coming up with a name, which displays the meaning even to those who haven't read the code. I want the tag to be optional in xml, meaning we only have to add it for the yields we want to remove. The standard code will have bools set to false if not mentioned in xml.

**Comment by raystuttgart** [comment]
Isn't this already solved, since the CityCenter plot Yield can now be actively chosen?

**Comment by Nightinggale** [comment]
It's not solved because while the player can choose the yield to produce, there is still a list of banned yields. I moved this to xml in the YieldCategory branch, making it much easier to configure (both no recompile and it's in just one location).

The issue is right now the following is banned and will never show up when choosing which yield to produce:

- lumber
- stone
- hemp
- horses
- cattle
- sheep

This issue asks the question why? And if we want to restrict, what should be disallow on colony plots?

**Comment by raystuttgart** [comment]
But well, that is good and makes sense.

Why should center City Plot Produce Livestock? 
(Or have you ever heard of Ranching been done in the middle of a City?)
-
How should it produce Lumber if all Forrest is removed?
-
Why should it produce stone which you only get on Peaks?

Summary:
I really see no problem that those Yields are not chooseable ...
We should not artifically create problems where there are none and then even code unrealistic stuff.

**Comment by Nightinggale** [comment]
> How should it produce Lumber if all Forrest is removed?

Easy, the plot will not produce lumber, hence no production. If somebody plants a forest or we implement #152 then lumber is missing. The same goes for stone.

What I'm puzzled about is hemp. Why can the colony grow cotton or tobacco, but not hemp?

**Comment by raystuttgart** [comment]
Hi Nightinggale, if you look into this function, it might answer your question:

bool CvPlayerAI::AI_isYieldForSale(YieldTypes eYield) const

**HEMP** is not a Yield to be sold.
(I considered it to be strategic for Rope or Sailcloth)

That **might be the reason** why you can not choose it on City Plot.

**Comment by raystuttgart** [comment]
I am really sorry, I coded a lot of stuff that I never explained to anybody and never documented because I thought that I would be "modding RaR forever" ...

I realize not that it was very stupid.

**Comment by Nightinggale** [comment]
> I am really sorry, I coded a lot of stuff that I never explained to anybody and never documented because I thought that I would be "modding RaR forever" ...
> 
> I realize not that it was very stupid.

I'm feeling that I'm guilty of the same issue even when I look at my own old code. It's sometimes obvious when working with it, but when you look at it much later or if it's from somebody else, then it's no longer obvious. That's why it's really hard to document correctly because overdoing comments is also bad. People tend to not read comments.

**Comment by raystuttgart** [comment]
YIELD_HEMP can now be selected on CenterCityPlot

**Comment by raystuttgart** [comment]
Hi guys,

can we close this issue?

- Hemp is fixed
- Livestock Yields don't make sense on CityCenterPlot
- Stone and Lumber don't make sense either

**Comment by Nightinggale** [comment]
> Stone and Lumber don't make sense either

I disagree with that statement. If a colony can produce ore, then why not stone?

We shouldn't restrict based on "the plot can't produce it anyway" because if the plot can't produce it, then it's disabled anyway. We should think "if the plot can produce the yield, should it be possible to produce on the colony plot?". I would go as far as to ban livestock and food (because food is always produced) and nothing else.

Yes you can produce lumber if there is a forest on the colony plot, but honestly it's not that great because you trade in some food for some lumber. You will get a better result from not having a forest and use the extra food on a lumberjack. This isn't about what is the best solution, but rather what the player should be allowed to do if the player wants to do so.

**Comment by raystuttgart** [comment]
Hm, ok.

If you want to correct that, it does not matter to me because you are right, it will probably not change anything ingame currently.
(You are absolutely right with "it is disabled" anyways - that was my point as well.)

I simply don't like **to code without purpose** (no bug or no feature) because it might some day have side effects as well.

Something like "Feature not removed when Founding City" we should discuss in Community Forum of course, if it is something you want to have.
It does definitely have an effect on gameplay.

Summary:
We basically agree.
I simply don't like coding / changing code if there is no (forseeable) reason to do it.

**Comment by Nightinggale** [comment]
I will close this because it's implemented just fine for 2.7.x. There is no need to change anything right now, which doesn't affect gameplay anyway. The question about stone and lumber is for finishing the YieldCategory branch, which is for 2.8.

**Comment by dcrookston** [comment]
Why doesn't livestock make sense?  A lot of people had grazing land for
livestock inside of a city back then.  If you've got room to grow some
cotton why can't you raise some cattle too?
